### PR TITLE
Warn for unreachable code after panic

### LINKED ISSUE #3023: Warn for unreachable code after todo and panic
Right now any argument given is simply ignored, leading to a bit confusing behavior.
While it may be syntactically possible to give an argument to the two keywords, they will mostly be ignored right now.

The tour teaches that the correct way to give a message is using the `as` keyword, but that is rather late in the tour and easily forgotten later, so it would be nice to give a hint and remind people of that gotcha.

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This PR closes #3023

I tried copying Rust's approach where it tries to understand if the previous expression will always panic and raise a warning accordingly. I'm not sure about the error message, let me know if you think it can be improved somehow!

### CODE REVIEW

**Review by lpil — COMMENTED** [review]
Lovely error messages! Looks so good!

One thing I don't like so much about this implementation is that it traverses the sub-tree for every single AST node, which is very wasteful. I think we could instead store whether the previously analysed expression is a panic on the ExprTyper in place of the bool that's used currently to prevent emitting multiple warnings. Then it's O(n) rather than O(n log n)

Let's remove the PanicKind too, saying the previous one panics is good enough I think.

**Comment by giacomocavalieri** [comment]
> I think we could instead store whether the previously analysed expression is a panic on the ExprTyper in place of the bool that's used currently to prevent emitting multiple warnings.

I'm not entirely sure I see how that could be implemented, so I would have to change every `infer_x` function to update a field in the typer? For cases that wouldn't work right away because I have to check that _all_ branches panic; I would have to type a branch, save the value of the field and go on with the others accumulating that result somewhere. Is this how it should work?

**Comment by lpil** [comment]
I think just the `infer` function and the ones for panic, anonymous functions, and case should be sufficient. In case I'd use a mutable variable as an accumulator to see if all clauses panic.

**Comment by giacomocavalieri** [comment]
> I think just the `infer` function and the ones for panic, anonymous functions, and case should be sufficient. In case I'd use a mutable variable as an accumulator to see if all clauses panic.

Thanks for the help! I should have fixed it now

File: compiler-core/src/type_/expression.rs [hunk 1]
```diff
@@ -185,6 +185,9 @@ pub enum ArgumentKind {
 pub(crate) struct ExprTyper<'a, 'b> {
     pub(crate) environment: &'a mut Environment<'b>,
 
+    pub(crate) previous_panics: bool,
+    pub(crate) already_warned_for_unreachable_code: bool,
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:189]: Some comments for these perhaps?

**Review by lpil — COMMENTED** [review]
Super nice!! I've left two small notes

File: compiler-core/src/type_/pipe.rs [hunk 2]
```diff
@@ -306,4 +324,37 @@ impl<'a, 'b, 'c> PipeTyper<'a, 'b, 'c> {
...
     }
+
+    fn warn_if_is_todo_or_panic(
+        &self,
+        call: &UntypedExpr,
+        first_location: SrcSpan,
+        previous_expression_location: Option<SrcSpan>,
+    ) {
+        let call_todo_or_panic = match call {
+            UntypedExpr::Todo { .. } => Some(TodoOrPanic::Todo),
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/type_/pipe.rs:335]: Oh yeah sorry I should have been clearer here, the addition is not related to this PR's feature but adds a warning if you use `panic` or `todo` as functions inside a pipeline just like with `panic(thing)` and `todo(thing)`, you can see the snapshot here <https://github.com/gleam-lang/gleam/pull/3148/files#diff-005eb040358d841482cd09562047007c76b266305bc4cbccf3d0ef0b80c415c9>

I was hands deep in the pipe typer then I noticed we didn't handle this case and thought I'd add it here directly without making a second PR, sorry!

[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/type_/pipe.rs:335]: Oh! Thank you, I understand now.

I didn't think we had any examples of people trying to attach a string to a `todo` or a `panic` using a pipeline line this. Is that a mistake folks are making?

File: compiler-core/src/type_/expression.rs (hunk 1 above)
Comment (giacomocavalieri) [reply compiler-core/src/type_/expression.rs:198]: Sure!

**Review by lpil — APPROVED** [review]
Delightful!! Thank you

**Pull request merged**